standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java

```diff
@@ -6672,12 +6674,26 @@ public GetOpenTxnsInfoResponse get_open_txns_info() throws TException {
 
     @Override
     public OpenTxnsResponse open_txns(OpenTxnRequest rqst) throws TException {
-      return getTxnHandler().openTxns(rqst);
+      OpenTxnsResponse response = getTxnHandler().openTxns(rqst);
+      List<Long> txnIds = response.getTxn_ids();
+      if (!listeners.isEmpty()) {
+        MetaStoreListenerNotifier.notifyEvent(listeners, EventType.OPEN_TXN,
+            new OpenTxnEvent(txnIds.iterator(), this));
+      }
+      if (!transactionalListeners.isEmpty()) {
```

Comment:
transactionalListeners should be notified within a txn boundary of metastore RDBMS. In this case, RDBMS open/commit happens within getTxnHandler().openTxns but we notify event afterwards. If notifyEvent fails, then we need to rollback the RDBMS txn.
Same comment for commit/abort as well.

Reply:
anyways we can not guarantee this as the logging of these notifications to RDBMS will happen in a different thread and its asynchronous I feel it need not be in the same txn scope.

Reply:
push txn listener to txn handler and add listener for abort and commit